dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/tasks/Ds3PutJob.java

```diff
@@ -35,306 +33,261 @@
+            job.withMetadata(new MetadataAccessImpl(fileMapper));
         }
+
+        addWaitingForChunkListener(totalJobSize, bucket + StringConstants.DOUBLE_SLASH + targetDir);
+        final AtomicLong totalSent = addDataTransferListener(totalJobSize);
+        job.attachObjectCompletedListener(obj -> updateGuiOnComplete(startJobDate, jobStartInstant, totalJobSize, totalSent, obj));
+
+        job.transfer(file -> FileChannel.open(PathUtil.resolveForSymbolic(fileMapper.get(file)), StandardOpenOption.READ));
+        waitForPermanentStorageTransfer(totalJobSize);
+        ParseJobInterruptionMap.removeJobID(jobInterruptionStore, this.getJobId().toString(), ds3Client.getConnectionDetails().getEndpoint(), deepStorageBrowserPresenter, loggingService);
     }
 
-    /**
-     * Create map of folders from the selected list
-     *
-     * @param directories      selected directories
-     * @param expandedPaths    expandedPaths
-     * @param partOfDirBuilder partOfDirBuilder
-     */
-    public ImmutableMap.Builder<String, Path> createFolderMap(final ImmutableList<Path> directories,
-                                                              final ImmutableMultimap.Builder<Path, Path> expandedPaths,
-                                                              final ImmutableSet.Builder<Path> partOfDirBuilder) {
-        final ImmutableMap.Builder<String, Path> folderMap = ImmutableMap.builder();
-        directories.forEach(path -> {
+    private static void buildMaps(final ImmutableMap.Builder<String, Path> fileMapBuilder, final ImmutableMap.Builder<String, Path> folderMapBuilder, final Pair<String, Path> pair, final LoggingService loggingService, final String targetDir) {
+        final String name = pair.getKey();
+        final Path path = pair.getValue();
+        if (hasNestedItems(path)) {
             try {
-                partOfDirBuilder.add(path);
-                expandedPaths.putAll(path, Files.walk(path).filter(child -> !Files.isDirectory(child)).collect(GuavaCollectors.immutableList()));
-                final String ds3ObjPath = getDs3ObjectPath(path, path, true, files.size(), directories.size());
-                final String ds3FileName = PathUtil.toDs3Path(targetDir, ds3ObjPath);
-                folderMap.put(ds3FileName, path);
+                Files.walk(path).filter(child -> !hasNestedItems(child)).map(p -> new Pair<>(targetDir + name + "/" + path.relativize(p).toString() + appendSlashWhenDirectory(p), p))
+                        .forEach(p -> folderMapBuilder.put(p.getKey(), p.getValue()));
             } catch (final IOException e) {
-                LOG.error("Failed to list files for directory : " + path.toString(), e);
-                loggingService.logMessage("Failed to list files for directory " + path.toString(), LogType.ERROR);
+                LOG.error("Unable to traverse provided path", e);
+                loggingService.logMessage("Tried to walk " + path.toString() + " but could not", LogType.ERROR);
```

Comment:
Can we get more information about the exception and pass that to the user, like a "we don't have permissions" or something like that?

Reply:
It looks like I can't eliminate the "don't know what went wrong" IOException case but it does look like I can check if it is a permissions issue and put up a different message.